### PR TITLE
MySql

### PR DESCRIPTION
**Pull request by srollinet**
Add support for MySql/MariaDb

### CODE REVIEW

**Comment by srollinet** [comment]
here it is :)

a few things I wasn't sure:
- Do you have a preference for base docker image versions? I took the lest precise current ones
  - mysql:8
  - mariadb:11
- I took MySqlConnector v0.49.2, it seems to be one of the most downloaded version. But it is probably possible to use an older version

**Comment by srollinet** [comment]
Seems build is failing because of SonarQube
`The format of the analysis property sonar.login= is invalid`

$SONAR_TOKEN was probably not set

File: README.md [hunk 1]
```diff
@@ -64,6 +64,7 @@ instances of common databases or anything else that can run in a Docker containe
 | MsSql Container      | [README](src/Container.Database.MsSql/README.md)      | [![MySql](https://img.shields.io/nuget/v/TestContainers.Container.Database.MsSql.svg)](https://www.nuget.org/packages/TestContainers.Container.Database.MsSql/)
 | PostgreSql Container | [README](src/Container.Database.PostgreSql/README.md) | [![PostgreSql](https://img.shields.io/nuget/v/TestContainers.Container.Database.PostgreSql.svg)](https://www.nuget.org/packages/TestContainers.Container.Database.PostgreSql/)
 | ArangoDb Container   | [README](src/Container.Database.ArangoDb/README.md)   | [![ArangoDb](https://img.shields.io/nuget/v/TestContainers.Container.Database.ArangoDb.svg)](https://www.nuget.org/packages/TestContainers.Container.Database.ArangoDb/)
+| Mysql Container      | [README](src/Container.Database.MySql/README.md)      | [![ArangoDb](https://img.shields.io/nuget/v/TestContainers.Container.Database.MySql.svg)](https://www.nuget.org/packages/TestContainers.Container.Database.MySql/)
```
Comment (isen-ng) [inline README.md]:
```suggestion
| Mysql Container      | [README](src/Container.Database.MySql/README.md)      | [![MySql](https://img.shields.io/nuget/v/TestContainers.Container.Database.MySql.svg)](https://www.nuget.org/packages/TestContainers.Container.Database.MySql/)
```

File: src/Container.Database.MySql/Container.Database.MySql.csproj [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    <SymbolPackageFormat>snupkg</SymbolPackageFormat>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <AssemblyName>TestContainers.Container.Database.MySql</AssemblyName>
+    <Description>DotNet port of testcontainers.org</Description>
+
+    <Version>1.4.9</Version>
+    <Authors>Isen Ng</Authors>
```
Comment (isen-ng) [inline src/Container.Database.MySql/Container.Database.MySql.csproj]: Would you like to add your name to the list of authors?
You can add it as a comma separated list.

**Comment by isen-ng** [comment]
>     * Do you have a preference for base docker image versions? I took the lest precise current ones
>       
>       * mysql:8
>       * mariadb:11

For the base image, let's stick with mysql. My gut feel is that this still more popular that mariadb... but its just my gut feeling.

>     * I took MySqlConnector v0.49.2, it seems to be one of the most downloaded version. But it is probably possible to use an older version

I think this should be fine, unless someone else complains. The license for this connector is MIT, so this is good. The license for the java version of mysqlconnector is GPL2, so for profit companies must take note, or switch to mariadb connector with is MIT.
For t

**Comment by isen-ng** [comment]
I'll figure out the sonar problem, don't worry about that!

**Comment by isen-ng** [comment]
Thank you for your PR!
Your work is very good!

**Comment by srollinet** [comment]
> For the base image, let's stick with mysql. My gut feel is that this still more popular that mariadb... but its just my gut feeling.

Yes mardiadb:11 is only for the integration tests, base image for the container is mysql:8

> Thank you for your PR!
> Your work is very good!

No problem, glad to help! Most of the code is yours with a few renames ;)
The code is really clean and easy to work with!

**Comment by srollinet** [comment]
I don't think the CI failure is related to that. It seems more like a random transient failure. Is it possible to re-run the CI?

As you can see, in the first build the MySql tests have been succesfully tested https://travis-ci.com/github/isen-ng/testcontainers-dotnet/builds/154670023#L956

And the failing build failed in an early stage, even before the MySql test execution. https://travis-ci.com/github/isen-ng/testcontainers-dotnet/builds/154982890#L508

> To be honest, I don't even know why this makes a difference. Based on my understanding of async and await, changing this from return to await is purely syntactic sugar but in this case, it somehow matters...

It is not only syntactic sugar. When the method is async, the compiler will generate a state machine and use it in your method to handle your async code.

See the difference between the two following methods

```c#
        public async Task InitializeAsync()
        {
            await Container.StartAsync();
        }

        public Task DisposeAsync()
        {
            return Container.StopAsync();
        }
```

You can see here all the code generated for `InitializeAsync()`

```c#
public abstract class PortBindingContainerFixture : IAsyncLifetime
{
	[CompilerGenerated]
	private sealed class <InitializeAsync>d__16 : IAsyncStateMachine
	{
		public int <>1__state;

		public AsyncTaskMethodBuilder <>t__builder;

		public PortBindingContainerFixture <>4__this;

		private TaskAwaiter <>u__1;

		private void MoveNext()
		{
			int num = <>1__state;
			try
			{
				TaskAwaiter awaiter;
				if (num != 0)
				{
					awaiter = <>4__this.Container.StartAsync().GetAwaiter();
					if (!awaiter.IsCompleted)
					{
						num = (<>1__state = 0);
						<>u__1 = awaiter;
						<InitializeAsync>d__16 stateMachine = this;
						<>t__builder.AwaitUnsafeOnCompleted(ref awaiter, ref stateMachine);
						return;
					}
				}
				else
				{
					awaiter = <>u__1;
					<>u__1 = default(TaskAwaiter);
					num = (<>1__state = -1);
				}
				awaiter.GetResult();
			}
			catch (Exception exception)
			{
				<>1__state = -2;
				<>t__builder.SetException(exception);
				return;
			}
			<>1__state = -2;
			<>t__builder.SetResult();
		}

		void IAsyncStateMachine.MoveNext()
		{
			//ILSpy generated this explicit interface implementation from .override directive in MoveNext
			this.MoveNext();
		}

		[DebuggerHidden]
		private void SetStateMachine(IAsyncStateMachine stateMachine)
		{
		}

		void IAsyncStateMachine.SetStateMachine(IAsyncStateMachine stateMachine)
		{
			//ILSpy generated this explicit interface implementation from .override directive in SetStateMachine
			this.SetStateMachine(stateMachine);
		}
	}

	...

	[AsyncStateMachine(typeof(<InitializeAsync>d__16))]
	[DebuggerStepThrough]
	public Task InitializeAsync()
	{
		<InitializeAsync>d__16 stateMachine = new <InitializeAsync>d__16();
		stateMachine.<>4__this = this;
		stateMachine.<>t__builder = AsyncTaskMethodBuilder.Create();
		stateMachine.<>1__state = -1;
		AsyncTaskMethodBuilder <>t__builder = stateMachine.<>t__builder;
		<>t__builder.Start(ref stateMachine);
		return stateMachine.<>t__builder.Task;
	}

	public Task DisposeAsync()
	{
		return Container.StopAsync();
	}
}
```

**Comment by isen-ng** [comment]
I have been retrying the build a few times everyday, hoping that it is a transient fault of the environment but it refuses to get past the same point.

To be honest, I really don't know what's going on with the build, however, changing `return` to `await` in previous manifestations of this problem has made the issue go away. That's why I am suggesting it for this time too.

**Comment by isen-ng** [comment]
It's stuck. Again.

**Comment by srollinet** [comment]
Is it possible to build the same commit again? https://travis-ci.com/github/isen-ng/testcontainers-dotnet/builds/154670023

To see if it also stucks now

**Comment by isen-ng** [comment]
Now that the other PR was built, this decides to take the status of the other PR and report a successful status..

Closing this since the other one was merged

**Pull request closed without merging**